File: frontend/js/account.js
// Page "Mon profil" : e-mail modifiable, nom/prenom renseignables une seule fois, identifiant fige.

const ACCOUNT_ERRORS = {
  invalid_email: "Adresse e-mail invalide.",
  invalid_name: "Le nom et le prénom ne doivent contenir que des lettres, espaces, apostrophes ou tirets.",
  identity_locked: "Ces informations ne peuvent plus être modifiées : contactez un administrateur."
};

const ACCOUNT_STATUS_LABELS = { active: "Actif", pending: "En attente de validation", disabled: "Désactivé" };

let accountProfile = null;

function accountEl(id) {
  return document.getElementById(id);
}

function accountEscape(value) {
  return String(value ?? "").replace(/[&<>"']/g, (c) => ({ "&": "&amp;", "<": "&lt;", ">": "&gt;", '"': "&quot;", "'": "&#39;" }[c]));
}

async function accountCsrf() {
  const response = await fetch("/api/csrf-token", { credentials: "same-origin" });
  return (await response.json()).token || "";
}

function renderAccount(profile) {
  accountProfile = profile;
  accountEl("acc_username").value = profile.username;
  [["first_name", "acc_first_name"], ["last_name", "acc_last_name"]].forEach(([key, id]) => {
    const input = accountEl(id);
    input.value = profile[key];
    input.readOnly = profile.locked[key];
  });
  accountEl("acc_email").value = profile.email;

  const editableNames = !profile.locked.first_name || !profile.locked.last_name;
  accountEl("acc_names_help").textContent = editableNames
    ? "Renseignable une seule fois : une fois enregistrés, seul un administrateur pourra modifier votre nom et votre prénom."
    : "Nom et prénom verrouillés : contactez un administrateur pour les corriger.";

  const rows = [
    ["Groupes", profile.groups.map((group) => group.label).join(", ") || "—"],
    ["Service", profile.service || "—"],
    ["Statut", ACCOUNT_STATUS_LABELS[profile.status] || profile.status],
    ["Compte créé le", profile.created_at ? new Date(profile.created_at).toLocaleDateString("fr-FR") : "—"]
  ];
  accountEl("accountAccess").innerHTML = rows.map(([label, value]) =>
    `<dt class="col-5 text-muted fw-normal">${accountEscape(label)}</dt><dd class="col-7">${accountEscape(value)}</dd>`).join("");
}

function showAccountMessage(level, text) {
  const box = accountEl("accountResult");
  box.className = `alert alert-${level} mt-3`;
  box.textContent = text;
}

async function loadAccount() {
  try {
    const response = await fetch("/api/account", { credentials: "same-origin", cache: "no-store" });
    if (!response.ok) throw new Error();
    renderAccount(await response.json());
  } catch (error) {
    showAccountMessage("danger", "Impossible de charger votre profil.");
  }
}

document.addEventListener("DOMContentLoaded", () => {
  const form = accountEl("accountForm");
  if (!form) return;
  accountEl("accountPasswordBtn").addEventListener("click", () => openPasswordChangeModal());

  form.addEventListener("submit", async (event) => {
    event.preventDefault();
    const errorBox = accountEl("accountError");
    errorBox.classList.add("d-none");
    const payload = { email: accountEl("acc_email").value.trim() };
    const settingNames = [];
    [["first_name", "acc_first_name"], ["last_name", "acc_last_name"]].forEach(([key, id]) => {
      const value = accountEl(id).value.trim();
      if (!accountProfile.locked[key] && value) {
        payload[key] = value;
        settingNames.push(key);
      }
    });
    if (settingNames.length && !(await askConfirm("Votre nom et votre prénom ne pourront plus être modifiés par vous après enregistrement. Confirmer ?", { confirmLabel: "Enregistrer définitivement" }))) {
      return;
    }
    const button = accountEl("accountSaveBtn");
    button.disabled = true;
    try {
      const response = await fetch("/api/account", {
        method: "PUT",
        credentials: "same-origin",
        headers: { "Content-Type": "application/json", "X-CSRF-Token": await accountCsrf() },
        body: JSON.stringify(payload)
      });
      const data = await response.json();
      if (!response.ok) {
        errorBox.textContent = ACCOUNT_ERRORS[data.error] || data.message || "Enregistrement impossible.";
        errorBox.classList.remove("d-none");
        return;
      }
      renderAccount(data);
      showAccountMessage("success", "Vos informations ont été enregistrées.");
    } catch (error) {
      errorBox.textContent = "Erreur de connexion au serveur.";
      errorBox.classList.remove("d-none");
    } finally {
      button.disabled = false;
    }
  });

  loadAccount();
});

File: backend/routes/account.py
"""Page et API « Mon compte » : l'utilisateur consulte et complete son propre profil."""
from flask import Blueprint, jsonify, send_from_directory

import account_rules
from auth import get_user_record, list_all_groups, login_required, rate_limit, update_user
from config import FRONTEND_DIR
from database import get_db
from flask import request, session
from models.audit import insert_app_log

bp = Blueprint("account", __name__)

_ERROR_MESSAGES = {
    "invalid_email": "Adresse e-mail invalide.",
    "invalid_name": "Nom ou prénom invalide.",
    "identity_locked": "L'identifiant, le nom et le prénom ne peuvent plus être modifiés : contactez un administrateur.",
}


def _profile(username):
    record = get_user_record(username)
    if not record:
        return None
    labels = {group["key"]: group.get("label") or group["key"] for group in list_all_groups()}
    return {
        "username": record["username"],
        "first_name": record.get("first_name") or "",
        "last_name": record.get("last_name") or "",
        "email": record.get("email") or "",
        "service": record.get("service") or "",
        "status": record.get("status") or "active",
        "created_at": record.get("created_at"),
        "groups": [{"key": key, "label": labels.get(key, key)} for key in record.get("groups", [])],
        # Champs deja renseignes : verrouilles pour l'utilisateur (seul un administrateur peut les changer).
        "locked": {key: bool(record.get(key)) for key in account_rules.SELF_EDITABLE_ONCE},
    }


@bp.route("/account.html")
@login_required
def account_page():
    return send_from_directory(FRONTEND_DIR, "account.html")


@bp.route("/api/account", methods=["GET"])
@login_required
def account_get():
    profile = _profile(session.get("user"))
    if not profile:
        return jsonify({"error": "not_found"}), 404
    return jsonify(profile)


@bp.route("/api/account", methods=["PUT"])
@login_required
@rate_limit(max_requests=20, window_seconds=60, scope="account_update")
def account_update():
    username = session.get("user")
    record = get_user_record(username)
    if not record:
        return jsonify({"error": "not_found"}), 404
    fields, error = account_rules.build_self_update(record, request.get_json(silent=True) or {})
    if error:
        return jsonify({"error": error, "message": _ERROR_MESSAGES.get(error, error)}), 403 if error == "identity_locked" else 400
    if fields and not update_user(username, **fields):
        return jsonify({"error": "update_failed"}), 500
    if fields:
        with get_db() as connection:
            insert_app_log(connection, "security", "account_self_update", "Profil modifié par l'utilisateur",
                           "user", username, {"fields": sorted(fields)}, actor=username)
    return jsonify(_profile(username))
